File: backend/app/api/endpoints/ground_station.py
from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session
from app.db.session import SessionLocal
from app.services.ground_station import (
    get_visible_now,
    get_next_passes,
    STATION,
    MIN_ELEVATION_DEG,
)
from datetime import datetime
from typing import Optional

router = APIRouter()


def get_db():
    db = SessionLocal()
    try:
        yield db
    finally:
        db.close()


@router.get("/info")
def station_info():
    """Return ground station metadata."""
    return {
        "name": STATION["name"],
        "latitude": STATION["lat_deg"],
        "longitude": STATION["lon_deg"],
        "altitude_m": STATION["alt_m"],
        "min_elevation_deg": MIN_ELEVATION_DEG,
    }


@router.get("/visibility/current")
def visibility_current(
    timestamp: Optional[datetime] = None,
    db: Session = Depends(get_db),
):
    """
    Return satellites currently visible from the ground station.
    Each entry includes norad_id, name, azimuth, elevation, range_km.
    """
    if timestamp is None:
        timestamp = datetime.utcnow()

    visible = get_visible_now(db, timestamp)
    return {
        "station": STATION["name"],
        "timestamp": timestamp.isoformat() + "Z",
        "count": len(visible),
        "satellites": visible,
    }


@router.get("/passes/next")
def passes_next(
    timestamp: Optional[datetime] = None,
    hours: float = Query(default=24.0, le=72.0, ge=1.0),
    db: Session = Depends(get_db),
):
    """
    Predict upcoming satellite passes within a time window.
    Each pass: norad_id, name, aos, los, duration_s, max_elevation, max_el_time.
    """
    if timestamp is None:
        timestamp = datetime.utcnow()

    passes = get_next_passes(db, timestamp, hours=hours)
    return {
        "station": STATION["name"],
        "window_start": timestamp.isoformat() + "Z",
        "window_hours": hours,
        "count": len(passes),
        "passes": passes,
    }

File: frontend/src/services/groundStation.ts
import axios from 'axios';

const API_URL = import.meta.env.VITE_API_URL || 'http://localhost:3001/api/v1';

// ── Types ───────────────────────────────────────────

export interface StationInfo {
    name: string;
    latitude: number;
    longitude: number;
    altitude_m: number;
    min_elevation_deg: number;
}

export interface VisibleSatellite {
    norad_id: number;
    name: string;
    azimuth: number;
    elevation: number;
    range_km: number;
}

export interface SatellitePass {
    norad_id: number;
    name: string;
    aos: string;
    los: string;
    duration_s: number;
    max_elevation: number;
    max_el_time: string;
}

export interface VisibilityResponse {
    station: string;
    timestamp: string;
    count: number;
    satellites: VisibleSatellite[];
}

export interface PassesResponse {
    station: string;
    window_start: string;
    window_hours: number;
    count: number;
    passes: SatellitePass[];
}

// ── API Calls ───────────────────────────────────────

export const fetchStationInfo = async (): Promise<StationInfo> => {
    const response = await axios.get(`${API_URL}/ground-station/info`);
    return response.data;
};

export const fetchCurrentVisibility = async (
    timestamp?: string
): Promise<VisibilityResponse> => {
    try {
        const params: any = {};
        if (timestamp) params.timestamp = timestamp;
        const response = await axios.get(`${API_URL}/ground-station/visibility/current`, { params });
        return response.data;
    } catch (error) {
        console.error("Failed to fetch visibility", error);
        return { station: "", timestamp: "", count: 0, satellites: [] };
    }
};

export const fetchNextPasses = async (
    timestamp?: string,
    hours: number = 24
): Promise<PassesResponse> => {
    try {
        const params: any = { hours };
        if (timestamp) params.timestamp = timestamp;
        const response = await axios.get(`${API_URL}/ground-station/passes/next`, { params });
        return response.data;
    } catch (error) {
        console.error("Failed to fetch passes", error);
        return { station: "", window_start: "", window_hours: hours, count: 0, passes: [] };
    }
};
